### PR TITLE
Feature/profile updater

### PR DESCRIPTION
**Pull request by AzizaAzhibekova**

### CODE REVIEW

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import us.peaksoft.gadgetarium.entity.*;
+import us.peaksoft.gadgetarium.dto.*;
+import us.peaksoft.gadgetarium.exception.NotFoundException;
+import us.peaksoft.gadgetarium.security.JwtService;
+import us.peaksoft.gadgetarium.repository.UserRepository;
+import us.peaksoft.gadgetarium.service.ProfileService;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/public/profile")
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:18]: remove the slash before the api and remove the public after the api

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+@Tag(name = "ProfileController", description = "API endpoints for managing profile")
+public class ProfileController {
+
+    private final UserRepository userRepository;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:22]: next line

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    private final JwtService jwtService;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:23]: next line

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        return profileService.updateProfile(profileRequest, user.getId());
+    }
+
+    @Operation(description = "Changing Profile password from request.")
+    @PutMapping("/password")
+    public SimpleResponse changePassword(@RequestHeader("Authorization") String authorizationHeader, @RequestBody ProfileChangePasswordRequest passwordRequest) {
+        String token = getTokenFromAuthorizationHeader(authorizationHeader);
+        User user = getUserFromToken(token);
+        return profileService.changePassword(passwordRequest,user.getId());
+    }
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:48]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileChangePasswordRequest.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.*;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+public class ProfileChangePasswordRequest {
+    private String oldPassword;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileChangePasswordRequest.java:11]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileChangePasswordRequest.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    private String NewPassword;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileChangePasswordRequest.java:12]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.*;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+public class ProfileRequest {
+    private String firstName;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java:11]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    private String lastName;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java:12]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    private String email;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java:13]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    private String address;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileRequest.java:14]: next line

File: src/main/java/us/peaksoft/gadgetarium/dto/ProfileResponse.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package us.peaksoft.gadgetarium.dto;
+
+import lombok.*;
+@Getter
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class ProfileResponse {
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/dto/ProfileResponse.java:9]: why did you write the variable with big letters

File: src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package us.peaksoft.gadgetarium.service;
+
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import us.peaksoft.gadgetarium.dto.*;
+@Service
+    public interface ProfileService {
+    SimpleResponse updateProfile(ProfileRequest profileRequest, Long id);
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java:8]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    SimpleResponse changePassword(ProfileChangePasswordRequest passwordRequest, Long id);
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java:9]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+import us.peaksoft.gadgetarium.exception.NotFoundException;
+import us.peaksoft.gadgetarium.repository.*;
+import us.peaksoft.gadgetarium.service.ProfileService;
+
+
+@Service
+@RequiredArgsConstructor
+public class ProfileServiceImpl implements ProfileService {
+
+    private final UserRepository userRepository;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:18]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    private final AddressRepository addressRepository;
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:19]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+    private boolean checkOldPassword(String email, String oldPassword) {
+        User user = userRepository.findByEmail(email).orElseThrow(() -> new NotFoundException("User is not found"));
+        String encodedPassword = user.getPassword();
+        return passwordEncoder.matches(oldPassword, encodedPassword);
+    }
+
+    private String PassToHash(CharSequence password) {
+        return passwordEncoder.encode(password);
+    }
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:30]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        Address address = user.getAddress();
+        if (address == null) { result = "User address not specified."; return result;}
+        String country = address.getCountryName();
+        String state = address.getStateName();
+        String city = address.getCityName();
+        String street = address.getStreetName();
+        String index = address.getPostalCode();
+        result = country+" "+state+" "+city+" "+" "+street+" "+index;
+        return result;
+    }
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:43]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+        user.setEmail(profileRequest.getEmail());
+        user.setPhoneNumber(profileRequest.getPhoneNumber());
+        userRepository.save(user);
+        update.setHttpStatus(HttpStatus.OK);
+        update.setMessage("Profile information for id["+user.getId()+"] updated");
+        return update;
+    }
+
+    public SimpleResponse changePassword (ProfileChangePasswordRequest passwordRequest, Long id) {
+
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:76]: remove line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+                changed.setHttpStatus(HttpStatus.BAD_REQUEST);
+                changed.setMessage("The old password was entered incorrectly.");
+                return changed;
+        }
+        user.setPassword(PassToHash(passwordRequest.getNewPassword()));
+        userRepository.save(user);
+        changed.setHttpStatus(HttpStatus.OK);
+        changed.setMessage("Password for id["+user.getId()+"] updated");
+        return changed;
+    }
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:94]: next line

File: src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    public ResponseEntity<?> getProfile(Long id) {
+
```
Comment (DilnazTolubaiKyzy) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/ProfileServiceImpl.java:97]: remove line

**Review by DilnazTolubaiKyzy — COMMENTED** [review]
viewed

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    @ExceptionHandler(ExpiredJwtException.class)
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:51]: Why handler is in Controller?

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    public ResponseEntity<SimpleResponse> handleExpiredJwtException() {
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:52]: Why handler is in Controller?

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+        SimpleResponse response = new SimpleResponse();
+        response.setHttpStatus(HttpStatus.BAD_REQUEST);
+        response.setMessage("Invalid token");
+        return new ResponseEntity<>(response, HttpStatus.BAD_REQUEST);
+    }
+
+    private String getTokenFromAuthorizationHeader(String authorizationHeader) {
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:59]: Same

File: src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
+            throw new IllegalArgumentException("Invalid authorization header");
+        }
+        return authorizationHeader.substring(7);
+    }
+
+    private User getUserFromToken(String token) {
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/controller/ProfileController.java:66]: Same

File: src/main/java/us/peaksoft/gadgetarium/repository/AddressRepository.java [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package us.peaksoft.gadgetarium.repository;
+
+import jakarta.transaction.Transactional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import us.peaksoft.gadgetarium.entity.Address;
+@Repository
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/repository/AddressRepository.java:7]: Next line pls

File: src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java [hunk 26]
```diff
@@ -0,0 +1,13 @@
+package us.peaksoft.gadgetarium.service;
+
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import us.peaksoft.gadgetarium.dto.*;
+@Service
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/service/ProfileService.java:6]: Pls refactor class

File: src/main/resources/application-dev.yml [hunk 27]
```diff
@@ -0,0 +1,26 @@
+spring:
```
Comment (i-Coderr) [inline src/main/resources/application-dev.yml:1]: Pls don't push your dev properties

**Review by i-Coderr — COMMENTED** [review]
Commented at 10:55

**Review by i-Coderr — APPROVED** [review]
Looks good!

**Pull request merged**